File: dairnet.py
from sklearn.covariance import EllipticEnvelope
import numpy as np
from sklearn.impute import SimpleImputer, KNNImputer
import pandas as pd
import pickle as pk
from sklearn import metrics

def indexin(y, x):
    index = np.argsort(x)
    sorted_x = x[index]
    sorted_index = np.searchsorted(sorted_x, y)
    # indices = index[sorted_index]
    # return indices
    yindex = np.take(index, sorted_index, mode="clip")
    mask = x[yindex] == y
    return yindex, mask

dir = "/weka/Projects/dairnet/code/united/examples/ChoulesAnon/data/anon"
dr = pd.read_csv(f"{dir}_bak/train_10min.csv")
X = dr.iloc[:, 3:27]
X = X.drop_duplicates(subset = list(X)[1:], keep=False)
X.describe().transpose()
X = X.dropna()
X = X.to_numpy()

outliers_fraction = 0.01
algorithm = EllipticEnvelope(contamination=outliers_fraction, random_state=42)
algorithm.fit(X)

y_pred = algorithm.predict(X)
np.mean(y_pred)
["#377eb8", "#ff7f00"]

dt = pd.read_csv(f"{dir}/test_system_anomaly_10min_0.2.csv")
y_timeidx = pk.load(open(f"{dir}/test_system_anomaly_10min_0.2_indices.pkl",'rb'))
y_idx, mask = indexin(y_timeidx, dt["time_idx"].to_numpy())
np.mean(mask)
Y = dt.iloc[:, 3:27]
Y = Y.drop_duplicates(subset = list(Y)[1:], keep=False)
Y.describe().transpose()
Y = Y.dropna()
Y = Y.to_numpy()
y_pred = algorithm.predict(Y)
np.mean(y_pred)
len(y_pred)
y_gt = np.zeros(len(y_pred))
y_gt[y_idx] = 1
np.mean(y_gt)
fpr, tpr, thresholds = metrics.roc_curve(y_gt, y_pred, pos_label=1)
metrics.auc(fpr, tpr)
np.mean(y_gt == y_pred)

Y[y_idx]
b, d = Y.shape
np.setdiff
_idx = np.setdiff1d(np.arange(b), y_idx)
Y

dy = pk.load(f"{dir}/test_system_anomaly_10min_indices.pkl")

# x = np.array([3, 5, 7, 1, 9, 8, 6, 6])
# y = np.array([2, 1, 5, 10, 100, 6])
# idx, mask = indexin(x, y)
# idx[mask]

File: old/dairnet.jl
using Flux, JLD2, DataFrames, CSV
include("lib/utils.jl")
include("imports.jl")

dir = "/weka/Projects/dairnet/code/united/examples/ChoulesAnon/data/anon/"
dr = CSV.read("$dir/train_10min.csv", DataFrame)
cols = names(dr)
jj = findall(contains("Sensor22"), cols)
d22 = dr[!, jj]
@> eachcol(d22) unique.() length.()

dr[!, 2]
dr[!, :group_id]
dr[!, :time_idx]
describe(dr)
@> dr[!, :group_id] unique

@unpack EllipticEnvelope = pyimport("sklearn.covariance")

outliers_fraction = 0.01
algorithm = EllipticEnvelope(contamination=outliers_fraction, random_state=42)
X = Array(d22)

import numpy as np
from sklearn.impute import KNNImputer
nan = np.nan
X = [[1, 2, nan], [3, 4, 3], [nan, 6, 5], [8, 8, 7]]
imputer = KNNImputer(n_neighbors=2, weights="uniform")
imputer.fit_transform(X)

algorithm.fit(X)
t1 = time.time()
plt.subplot(len(datasets), len(anomaly_algorithms), plot_num)
y_pred = algorithm.fit(X).predict(X)
Z = algorithm.predict(np.c_[xx.ravel(), yy.ravel()])
Z = Z.reshape(xx.shape)
plt.contour(xx, yy, Z, levels=[0], linewidths=2, colors="black")

